engine/common/synchronization/engine.go

```diff
@@ -284,7 +284,10 @@ func (e *Engine) processAvailableResponses() {
 // onSyncResponse processes a synchronization response.
 func (e *Engine) onSyncResponse(originID flow.Identifier, res *messages.SyncResponse) {
 	e.log.Debug().Str("origin_id", originID.String()).Msg("received sync response")
-	final := e.finalizedHeader.Get()
+	final, err := e.state.Final().Head()
```

Comment:
I explained my reasoning in detail above, for why I disagree with this line of thinking: https://github.com/onflow/flow-go/pull/4187#issuecomment-1506112592. 

We have also discussed this last week and I suggested a rule of thumb for where to draw the line regarding when information should be communicated via an event:
> - An event should represent a significant state change in the system (eg. a new block has been finalized)
> - A component should subscribe to an event if it needs to perform business logic as a result of the state change

This rule of thumb:
- Forbids polling for state changes, because we want to encourage event-driven designs
- Discourages every component maintaining their own local copy of the subset of the global state they need to read, because doing so introduces a lot of complexity for little benefit in most cases

An example of following this rule is: a verification node subscribes to `OnBlockProcessable` and begins verifying a chunk it is assigned to in the newly processable block

An example of not following this rule is: an engine sometimes needs to know the latest value of variable `X`, **at times that are not related to when `X` changes**, so it subscribes to `OnXChanged` and updates a local copy of `X` every time it consumes the event. 

_Only_ updating a local cache of the latest finalized block, when consuming an `OnBlockFinalized` event, is for the same reason an example of not following this rule. As we've discussed, there are potentially performance reasons for having multiple local caches of the latest finalized block in particular, because it is commonly accessed. The business logic _for that cache_ must respond to `OnBlockFinalized` events in order to maintain the cache. The business logic _for the component using the cache_ may need to do this, or it may not. 

> * On the main path of the logic, we are polling the information from a central source (potentially many times the same value, before any update occurs)

There is a difference between polling and reading data. Polling is when we repeatedly query a value, _so that we can do something when that value changes_. Here, we read the data because we want to know it's current value -- we don't care when it changes. 

> * no event-driven pattern, no usage of `FinalizationActor`

We do not use the FinalizationActor in this engine because this engine _does not perform business logic in response to blocks being finalized_. This engine largely performs business logic in response to messages from other nodes (which are treated similarly to internal events). In this case, we are responding to the `SyncResponse` by enqueuing new heights which should be queried. When we respond to the `SyncResponse`, we need to know the current value for the latest finalized block (so we read it), but we don't care when it changes (so we don't subscribe to events describing when it changes).

> Would it very hard to switch to the FinalizationActor here?

It would not be hard to use the `FinalizationActor` here, but it would be a poor example for when it is appropriate to use.